src/Sphere.jl

```diff
@@ -2,10 +2,37 @@ struct Sphere{T} <: Manifold where {T}
     shape::T
 end
 
-project_tangent!(S::Sphere,w,x,v) = (w .= w - dot(x,v)*x)
-distance(S::Sphere,x,y) = norm(x-y) # FIXME
-exp!(S::Sphere,y,x,v) = retract!(S,y,x,v) # FIXME
-log!(S::Sphere,v,x,y) = project_tangent!(S,v,x,y-x) # FIXME
+isapprox(S::Sphere, x, y; kwargs...) = isapprox(x, y; kwargs...)
+project_tangent!(S::Sphere, w, x, v) = (w .= w .- dot(x, v).*x)
+distance(S::Sphere, x, y) = acos(dot(x, y))
+
+function exp!(S::Sphere, y, x, v)
+    nv = norm(S, x, v)
+    if nv ≈ 0.0
+        y .= x
+    else
+        y .= cos(nv).*x .+ (sin(nv)/nv).*v
+    end
+    return y
+end
+
+function log!(S::Sphere, v, x, y)
+    θ = acos(dot(x, y))
+    if θ ≈ 0.0
+        zero_tangent_vector!(S, v, x)
+    else
+        v .= (θ/sin(θ)) .* (y .- cos(θ).*x)
+    end
+    return v
+end
+
 dimension(S::Sphere) = sum(S.shape)
 random_point(S::Sphere) = (x = randn(S.shape); x / norm(x))
-random_tangent_vector(S::Sphere, x) = (v = randn(S.shape); project_tangent!(v,x,v))
+
+function random_tangent_vector(S::Sphere, x)
+    v = randn(S.shape)
```

Comment:
Hm, we might have to parametrize on a float type. How about we parametrize `Sphere` on Tfloat, and add convenience constructors like `Sphere(x) = Sphere{eltype(x)}(size(x))`?

Reply:
To be honest, I'd prefer to have different types for point and tangent vector distributions and extend `rand` (see #3 for a discussion)

Reply:
OK; then we also extend `zero` and have no need for storing types or shapes in the manifold structure?

Reply:
Do you mean renaming `zero_tangent_vector` to `zero`? `Sphere` will probably only have to store dimension (to implement `dimension`). I don't know if we should do this by a type parameter or a field.

Reply:
Yes, it'd be natural to extend `zero`. But I don't know what to do about point vs tangent vector. Same goes for random.

Seems overkill to store dimension just to be able to print it back. If we really want dimension we can ask the user to give us a point: `dimension(S::Sphere,x) = length(x)-1`.

Reply:
I'm not sure if it's clear enough that `zero(M, x)` returns a zero tangent vector. Why not zero element of a different vector bundle at `x`? Random points and tangent vectors are a separate issue, I'll have to think about it a bit more.

`dimension(M, x)` looks weird, all points on a manifold have the same dimension. A good use case for `dimension` would be nice to determine which option is better.

Reply:
Is there a particular reason that we store dimension as a field instead of as a parameter? The latter enables dispatch on dimension. Of course, this may be moot if we use specialized vector types.

Do we want to also provide spheres of non-unit radius?